File: app/src/main/java/com/microjet/airqi2/GestureLock/DefaultPatternCheckingActivity.java
package com.microjet.airqi2.GestureLock;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.TextView;

import com.github.ihsg.patternlocker.OnPatternChangeListener;
import com.github.ihsg.patternlocker.PatternIndicatorView;
import com.github.ihsg.patternlocker.PatternLockerView;
import com.microjet.airqi2.AirMapActivity;
import com.microjet.airqi2.Definition.SavePreferences;
import com.microjet.airqi2.R;
import com.microjet.airqi2.SettingActivity;

import java.util.List;


public class DefaultPatternCheckingActivity extends AppCompatActivity {
    private PatternIndicatorView patternIndicatorView;
    private TextView textMsg;
    private PatternHelper patternHelper;

    private static int actionMode = 0;

    public static final int START_ACTION_MODE_NORMAL = 0;
    public static final int START_ACTION_MODE_DISABLE = 1;
    public static final int START_ACTION_MODE_CHANGE_PASSWOPRD = 2;

    public static void startAction(Context context, int mode) {
        Intent intent = new Intent(context, DefaultPatternCheckingActivity.class);
        context.startActivity(intent);

        actionMode = mode;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_default_pattern_checking);

        this.patternIndicatorView = findViewById(R.id.pattern_indicator_view);
        PatternLockerView patternLockerView = findViewById(R.id.pattern_lock_view);
        this.textMsg = findViewById(R.id.text_msg);

        patternLockerView.setOnPatternChangedListener(new OnPatternChangeListener() {
            @Override
            public void onStart(PatternLockerView view) {
            }

            @Override
            public void onChange(PatternLockerView view, List<Integer> hitList) {
            }

            @Override
            public void onComplete(PatternLockerView view, List<Integer> hitList) {
                boolean isError = !isPatternOk(hitList);
                view.updateStatus(isError);
                patternIndicatorView.updateState(hitList, isError);
                updateMsg();

                if(!isError) {
                    switch (actionMode) {
                        case START_ACTION_MODE_DISABLE:
                            SharedPreferences share = getSharedPreferences(SavePreferences.SETTING_KEY, Context.MODE_PRIVATE);
                            share.edit().putBoolean(SavePreferences.SETTING_MAP_PRIVACY, false).apply();
                            break;

                        case START_ACTION_MODE_NORMAL:
                            callCompletePage();
                            break;

                        case START_ACTION_MODE_CHANGE_PASSWOPRD:
                            DefaultPatternSettingActivity.startAction(DefaultPatternCheckingActivity.this);
                            break;
                    }
                }
                Log.e("Pattern", "Error: " + isError);
            }

            @Override
            public void onClear(PatternLockerView view) {
                finishIfNeeded();
            }
        });

        this.textMsg.setText(getResources().getText(R.string.text_draw_pattern));
        this.patternHelper = new PatternHelper();
    }

    private boolean isPatternOk(List<Integer> hitList) {
        this.patternHelper.validateForChecking(hitList);
        return this.patternHelper.isOk();
    }

    private void updateMsg() {
        this.textMsg.setText(this.patternHelper.getMessage());
        this.textMsg.setTextColor(this.patternHelper.isOk() ?
                getResources().getColor(R.color.colorPrimary) :
                getResources().getColor(R.color.colorAccent));
    }

    private void finishIfNeeded() {
        if (this.patternHelper.isFinish()) {
            finish();
        }
    }

    private void callCompletePage() {
        Intent i = new Intent(this, AirMapActivity.class);
        startActivity(i);
    }
}

File: app/src/main/java/com/microjet/airqi2/GestureLock/DefaultPatternCheckingActivity.kt
package com.microjet.airqi2.GestureLock

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import com.github.ihsg.patternlocker.OnPatternChangeListener
import com.github.ihsg.patternlocker.PatternLockerView
import com.microjet.airqi2.AirMapActivity
import com.microjet.airqi2.GoldenMapActivity
import com.microjet.airqi2.PrefObjects
import com.microjet.airqi2.R
import kotlinx.android.synthetic.main.activity_default_pattern_checking.*
import java.util.*


class DefaultPatternCheckingActivity : AppCompatActivity() {

    private var patternHelper: PatternHelper? = null

    private lateinit var myPref: PrefObjects

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_default_pattern_checking)

        myPref = PrefObjects(this)

        val pref = getSharedPreferences("TOKEN", Context.MODE_PRIVATE)
        val userPW = pref.getString("LoginPassword", "")

        if(userPW == "") {
            btnUsePW.visibility = View.GONE
        } else {
            btnUsePW.visibility = View.VISIBLE
        }

        btnUsePW.setOnClickListener {
            UsePasswordActivity.startAction(this@DefaultPatternCheckingActivity, actionMode)
            finish()
        }

        patternLockerView.setOnPatternChangedListener(object : OnPatternChangeListener {
            override fun onStart(view: PatternLockerView) {}

            override fun onChange(view: PatternLockerView, hitList: List<Int>) {}

            override fun onComplete(view: PatternLockerView, hitList: List<Int>) {
                val isError = !isPatternOk(hitList)
                view.updateStatus(isError)
                //patternIndicatorView!!.updateState(hitList, isError)
                updateMsg()

                if (!isError) {
                    when (actionMode) {
                        START_ACTION_MODE_DISABLE -> {
                            myPref.setSharePreferencePrivacy(false)
                        }

                        START_ACTION_MODE_NORMAL -> callCompletePage()

                        START_ACTION_MODE_CHANGE_PASSWOPRD -> DefaultPatternSettingActivity.startAction(this@DefaultPatternCheckingActivity)
                    }
                }
                Log.e("Pattern", "Error: $isError")
            }

            override fun onClear(view: PatternLockerView) {
                finishIfNeeded()
            }
        })

        text_msg.text = resources.getText(R.string.text_draw_pattern)
        this.patternHelper = PatternHelper()
    }

    private fun isPatternOk(hitList: List<Int>): Boolean {
        this.patternHelper!!.validateForChecking(hitList)
        return this.patternHelper!!.isOk
    }

    private fun updateMsg() {
        text_msg.text = this.patternHelper!!.message
        text_msg.setTextColor(if (this.patternHelper!!.isOk)
            ContextCompat.getColor(this, R.color.Main_textResult_Good)
        else
            ContextCompat.getColor(this, R.color.Main_textResult_Bad))
    }

    private fun finishIfNeeded() {
        if (this.patternHelper!!.isFinish) {
            finish()
        }
    }

    private fun callCompletePage() {
        val mLang = Locale.getDefault().language + "-" + Locale.getDefault().country
        val i: Intent? = Intent(this, if (mLang == "zh-CN") {
            GoldenMapActivity::class.java
        } else {
            AirMapActivity::class.java
        })
        startActivity(i)
    }

    companion object {

        private var actionMode = 0

        const val START_ACTION_MODE_NORMAL = 0
        const val START_ACTION_MODE_DISABLE = 1
        const val START_ACTION_MODE_CHANGE_PASSWOPRD = 2

        fun startAction(context: Context, mode: Int) {
            val intent = Intent(context, DefaultPatternCheckingActivity::class.java)
            context.startActivity(intent)

            actionMode = mode
        }
    }
}
